README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# BMRSr <a href='https://bmrsr.arawles.co.uk'><img src='man/figures/logo.png' align="right" height="139" /></a>

[![CRAN
status](https://www.r-pkg.org/badges/version/BMRSr)](https://cran.r-project.org/package=BMRSr)
[![R-CMD-Check](https://github.com/ARawles/BMRSr/workflows/R-CMD-check/badge.svg)](https://github.com/ARawles/BMRSr/actions)
[![Coverage
status](https://codecov.io/gh/ARawles/BMRSr/branch/master/graph/badge.svg)](https://codecov.io/github/ARawles/BMRSr?branch=master)
[![](https://cranlogs.r-pkg.org/badges/BMRSr)](https://cran.r-project.org/package=BMRSr)

## Overview

The goal of BMRSr is to provide wrapper functions to make use of the
[Balancing Mechanism Reporting System
API](https://www.elexon.co.uk/documents/training-guidance/bsc-guidance-notes/bmrs-api-and-data-push-user-guide-2/)
easier for R users. <br> The [Balancing Mechanism Reporting
System](https://bmreports.com/bmrs/?q=eds/main) (or BMRS) is a platform
for energy-related data in the UK. Here, data relating to energy
generation, demand, transmission and balancing services can all be
accessed for free. The BMRS is therefore a platform for enthusiasts,
academics and professionals to extract energy data for analysis and
modelling. Also associated with BMRS is an API which allows for access
to all the data items currently available through the UI. <br> This R
package contains functions to help build the API requests, send them,
retrieve the requested data and parse it.

## Installation

You can install BMRSr from CRAN with:

``` r
install.packages("BMRSr")
```

You can install the development version of BMRSr from
[GitHub](https://github.com/ARawles/BMRSr) with:

``` r
devtools::install_github("BMRSr")
```

**You’ll also need an API key. You can get this by registering on the
[ELEXON portal](https://www.elexonportal.co.uk).**

## Usage

BMRSr contains functions that can be split into 4 main categories:

  - Build
  - Send & Receive
  - Parse
  - Utility

### End-to-End

To perform a complete API request (build the call, send and receive the
data, and then parse it), use the `full_request()` function.

``` r
full_request(data_item = "B1720",
             api_key = "test",
             settlement_date = "12 Jun 2018",
             period = "1",
             service_type = "csv")
```

### Build

These functions build the URL for the API request. The main function
`build_call()` is the one you’ll likely be calling, but all this does is
call the appropriate `build_x_call()` function for the data item you’ve
requested. For example:

``` r
build_call(data_item = "B1720",
           api_key = "12345",
           settlement_date = "1 Jan 2018",
           period = "1",
           service_type = "csv")
#> $url
#> [1] "https://api.bmreports.com/BMRS/B1720/v1?APIKey=12345&SettlementDate=2018-01-01&Period=1&ServiceType=csv"
#> 
#> $service_type
#> [1] "csv"
#> 
#> $data_item
#> [1] "B1720"
```

actually calls

``` r
build_b_call(data_item = "B1720",
             api_key = "12345",
             settlement_date = "1 Jan 2018",
             period = "1",
             service_type = "csv")
#> $url
#> [1] "https://api.bmreports.com/BMRS/B1720/v1?APIKey=12345&SettlementDate=2018-01-01&Period=1&ServiceType=csv"
#> 
#> $service_type
#> [1] "csv"
#> 
#> $data_item
#> [1] "B1720"
```

These functions return a list of three items:

  - the url as a character string (`$url`)
  - the service type/return format (`$service_type`)
  - the data item (`$data_item`)

The input parameters you provide will be checked against those that are
valid for the data item you are requesting, however there is no check on
whether you have provided (at least) the required parameters for the
data item.

To see all the allowed input parameters for each type (not each data
item), use `?build_[type]_call`.

To see which input parameter you need for your data item, use
`get_parameters(your_data_item)`.

### Send & Receive

This function - `send_request()` - sends the provided URL to the API and
returns a response() object with the added attribute of data\_item\_type
(one of “B Flow”, “Remit”, or “Legacy”). Config options can also be
supplied via the config\_options parameter as a named list, that will be
passed to the `httr::GET()` function (implemented primarily for proxies
and the like).

This function can be used with a premade url, however the user will also
have to respecify the data item and service type (format) from the URL:

``` r
send_request(
  list(
    url = "https://api.bmreports.com/BMRS/B1720/v1?APIKey=12345&SettlementDate=2018-01-01&Period=1&ServiceType=csv", 
                  data_item = "B1720", service_type = "csv")
)
```

### Parse

This function - `parse_response()` - takes the response() object
returned from the send\_request() function, and parses the response base
on the service\_type parameter (whether it was “csv” or “xml”). CSVs
return tibbles, and XMLs return lists. The returned CSVs from many of
the calls contain unnecessary or incorrect data, so this parsing
function will remove that data before returning a corrected response.

``` r
parse_response(
  send_request(
    list(
      url = "https://api.bmreports.com/BMRS/B1720/v1?APIKey=12345&SettlementDate=2018-01-01&Period=1&ServiceType=csv", 
                    data_item = "B1720", service_type = "csv")
  )
)
```

### Utility

These functions support the functionality of the previous 3 groups of
functions:

  - `get_function()` which returns the appropriate `build_x_call()`
    function needed for the `build_call()` function.
  - `check_data_item()` ensures that the request is for a valid data
    item.
  - `get_parameters()` returns a list with the allowed input parameters
    for the supplied data item.
  - `clean_date_columns()` reformats date/time/datetime columns based on
    their column names.
  - `get_data_items()` returns all valid data items.
  - `get_column_names()` retrieves the column headings for a particular
    data item (Legacy only as B flow responses already have column
    headings).
  - `get_data_item_type()` return the data item *type* of a particular
    data item. This is only really relevant to know which
    `build_x_call()` to use.

## Full example

Here’s a full example, using the package to return generation by fuel
type data and then plotting it using ggplot2.

``` r

api <- "your_api_key_goes_here"


#We're requesting the FUELINST data item here.
#A full list of all the data items can be returned using the get_data_items() function

#Find out which parameters we need to provide for the data item we've chosen...
get_parameters("FUELINST")
#> [1] "from_datetime" "to_datetime"
```

``` r
#Perform the full request and ask for a CSV back
gen_data <- full_request(data_item = "FUELINST",
                         api_key = api,
                         from_datetime = "2019-07-01 00:00:00",
                         to_datetime = "2019-07-03 00:00:00",
                         service_type = "csv")
```

``` r
#Load the libraries for a bit more cleaning and then plotting...
library(ggplot2, quietly = TRUE, warn.conflicts = FALSE)
library(dplyr, quietly = TRUE, warn.conflicts = FALSE)

head(gen_data)
#> # A tibble: 6 x 19
#>   record_type settlement_date settlement_peri~ spot_time            ccgt   oil
#>   <chr>       <date>                     <dbl> <dttm>              <dbl> <dbl>
#> 1 FUELINST    2019-07-01                     2 2019-07-01 00:00:00  4101     0
#> 2 FUELINST    2019-07-01                     3 2019-07-01 00:05:00  4354     0
#> 3 FUELINST    2019-07-01                     3 2019-07-01 00:10:00  4388     0
#> 4 FUELINST    2019-07-01                     3 2019-07-01 00:15:00  4290     0
#> 5 FUELINST    2019-07-01                     3 2019-07-01 00:20:00  4237     0
#> 6 FUELINST    2019-07-01                     3 2019-07-01 00:25:00  4176     0
#> # ... with 13 more variables: coal <dbl>, nuclear <dbl>, wind <dbl>, ps <dbl>,
#> #   npshyd <dbl>, ocgt <dbl>, other <dbl>, intfr <dbl>, intirl <dbl>,
#> #   intned <dbl>, intew <dbl>, biomass <dbl>, intnem <dbl>

#Change the fuel types from columns to a grouping
gen_data <- gen_data %>%
  dplyr::mutate(settlement_period = as.factor(settlement_period)) %>%
  tidyr::gather(key = "fuel_type", value = "generation_mw", ccgt:intnem)



#Make a line graph of the different generation types
ggplot2::ggplot(data = gen_data, aes(x = spot_time, y = generation_mw, colour = fuel_type)) +
  geom_line()
```

<img src="man/figures/README-unnamed-chunk-9-1.png" width="100%" />


NEWS.md
# BMRSr 1.0.3.9000

* Just use ISO date and datetime formats across the board instead of trying to accept lots of different formats
    + This just leads to dates being formatted incorrectly and not working
* Fixed the column renaming for the 'FUELINST' data item

# BMRSr 1.0.3

* Added better error handling around the sending and parsing of requests as per CRAN policies

# BMRSr 1.0.2.9000

* Fixed issue where EOF files were showing warnings when being parsed because the EOF wasn't being removed.
* Added check of data item to ensure that the data item is present for the provided version

# BMRSr 1.0.2

* Third CRAN release
* Changes documented in 1.0.1 - 1.0.2 updates

# BMRSr 1.0.1.9003

* Added automatic cleaning of all parameters passed to the `build()` functions
* Added `rlang` and `purrr` dependencies

# BMRSr 1.0.1.9002

* Expanded changes from `build_b_call()` to other `build` functions
* Allowed for additional arbitrary parameters to the build calls (with warnings or without)

# BMRSr 1.0.1.9001 

*  Changes to build_b_call() function
*  Collect query parameters using list input_query
*  Use `httr::modify_url()` to build query
*  Shorten long lines
*  Added test 'Typical Request' to test_that
*  Other cosmetic changes

# BMRSr 1.0.1.9000

* Changes to the API return values mean that the returned .csv files do not always have the same structure (#15).
    + This bred a change to the parser to be more dynamic and allow for the different structures.
    + This initial change has been made but further changes will need to be made to account for more file structures.

# BMRSr 1.0.1

* New CRAN release
* Bug fixes documented in versions 1.0.0 - 1.0.1
* New documentation website hosted at [bmrsr.arawles.co.uk](https://bmrsr.arawles.co.uk)
* Updated links to BMRS documentation

# BMRSr 1.0.0.9004

* Restructured the getting data items/data items types functions to pull from a single paired list of data items and types

# BMRSr 1.0.0.9002/3

* Fixed lingering typo issues

# BMRSr 1.0.0.9001

* Added `rename` parameter to `parse_response()` to avoid attempting to rename columns
* Fixed typo in `get_data_items()` list (DISBSAD was incorrectly spelt DISBAD) (#13)

# BMRSr 1.0.0

* Minor changes to prepare for CRAN publication

# BMRSr 0.1.3

* Added NEWS.md
* Included packages used in example in README.md in Suggests
* Updated documentation of included dataset to include description of every column
* Fixed versioning scheme
* Changed license to GPL (>= 2)

# BMRSr 0.1.2.001

* Fixed column types in example dataset (previously as characters, now as date/datetime depending on column)
* Added a new utility function `get_data_item_type()` to return the data item type ("B Flow", "REMIT" or "Legacy") of a data item
* Updated README to update changes
* Added first vignette - Using_BMRSr - which describes an end-to-end process and gives an in-depth example of how to use the package.
* Implemented a try statement into the column cleaning function to no longer error and stop if column conversion fails

# BMRSr 0.1.1.9019

* Fixed some incorrect column names returned by `get_column_names()`

# BMRSr 0.1.1.9016

* settlement_period and period parameters now accept "*" value to return all Settlement Periods


R/jim_build.R
# jim_build (wip:  version of build_b_call())
# Functions to build API calls


#' Create an API call for B-data flows
#'
#' @param data_item character string; the id of the B flow
#' @param api_key character string; api key retrieved from the Elexon portal
#'
#' settlement date (automatically cleaned by format_date)
#' @param settlement_date character string;
#' @param period character string; settlement period
#' @param year character string; year
#' @param month character string; month
#' @param week character string; week
#' @param process_type character string; process type
#' @param start_time character string; start time
#' @param end_time character string; end time
#' @param start_date character string; start date
#' @param end_date character string; end date
#' @param service_type character string; file format (csv or xml)
#'
#' version of the api to use (currently on v1)
#' @param api_version character string; 
#
#' @return list; created url for the call, service type and data item
#' @family call-building functions
#' @export
#' @examples
#' \dontrun{
#'     build_b_call(data_item = "B1730", 
#'     api_key = "12345", settlement_date = "14-12-2016")
#'
#'     build_b_call(data_item = "B1510", 
#'     api_key = "12345", start_date = "01 Jan 2019",
#'     start_time = "00:00:00", end_date = "02 Jan 2019", 
#'     end_time = "24:00:00", service_type = "csv")
#' }
#'
build_b_call <- function(data_item, api_key, 
        settlement_date = NULL, period = NULL,
        year = NULL, month = NULL, week = NULL, 
        process_type = NULL, start_time = NULL,
        end_time = NULL, start_date = NULL, end_date = NULL, 
        service_type = "csv", api_version = "v1") {

  request <- list()
  request$url <- paste0("https://api.bmreports.com/BMRS/", 
                       data_item, "/", api_version, "?APIKey=", api_key)

  check_data_item(data_item, "B Flow")

  if (!is.null(settlement_date)) {
    request$url <- paste0(request$url, "&SettlementDate=", 
                         format_date(settlement_date))
  }

  if (!is.null(period)){
    if (period <= 0 | period > 50){
      if (period != "*"){
      stop("invalid period value")
      }
    }
    request$url = paste0(request$url, "&Period=", as.character(period))
  }
  if (!is.null(process_type)){
    request$url = paste0(request$url, "&ProcessType", process_type)
  }
  if (!is.null(year)){
    request$url = paste0(request$url, "&Year=", year)
  }
  if (!is.null(month)){
    request$url = paste0(request$url, "&Month=", format_month(month))
  }
  if (!is.null(week)){
    request$url = paste0(request$url, "&Week=", week)
  }
  if (!is.null(start_date)){
    request$url = paste0(request$url, "&StartDate=", format_date(start_date))
  }
  if (!is.null(end_date)){
    request$url = paste0(request$url, "&EndDate=", format_date(end_date))
  }
  if (!is.null(start_time)){
    request$url = paste0(request$url, "&StartTime=", format_time(start_time))
  }
  if (!is.null(end_time)){
    request$url = paste0(request$url, "&EndTime=", format_time(end_time))
  }
  request$url = paste0(request$url, "&ServiceType=", service_type)
  request$service_type <- service_type
  request$data_item <- data_item
  return(request)
}

#' Create an API call for REMIT flows
#'
#' @param data_item character string; the id of the REMIT flow
#' @param api_key character string; api key retrieved from the Elexon portal
#' @param event_start character string; event start (automatically cleaned by format_date)
#' @param event_end character string; event end (automatically cleaned by format_date)
#' @param publication_from character string; publication from (automatically cleaned by format_date)
#' @param publication_to character string; publication to (automatically cleaned by format_date)
#' @param participant_id character string; participant id
#' @param asset_id character string; asset id
#' @param event_type character string; event type
#' @param fuel_type character string; fuel type
#' @param message_type character string; message type
#' @param message_id character string; message id
#' @param unavailability_type character string; unavailability type
#' @param active_flag character string; active flag
#' @param sequence_id character string; sequence id
#' @param service_type character string; file format (csv or xml)
#' @param api_version character string; version of the api to use (currently on v1)
#' @return list; created url for the call, service type and data item
#' @family call-building functions
#' @examples
#' build_remit_call(data_item = "MessageListRetrieval", api_key = "12345",
#' event_start = "14-12-2016", event_end = "15-12-2016")
#' build_remit_call(data_item = "MessageDetailRetrieval", api_key = "12345",
#' participant_id = 21, service_type = "xml")
#' @export

build_remit_call <- function(data_item, api_key, event_start = NULL, event_end = NULL, publication_from = NULL, publication_to = NULL,
                             participant_id = NULL, asset_id =  NULL, event_type = NULL, fuel_type = NULL, message_type = NULL, message_id = NULL,
                             unavailability_type =  NULL, active_flag = NULL, sequence_id = NULL, service_type = "xml", api_version = "v1"){
  request <- list()
  check_data_item(data_item, "REMIT")
  request$url = paste0("https://api.bmreports.com/BMRS/", data_item, "/", api_version, "?APIKey=", api_key)
  if (service_type == "csv"){
    warning("Remit files cannot be returned as .csv - file will be returned as xml")
    service_type <- "xml"
  }
  if (!is.null(event_start)) {
    request$url = paste0(request$url, "&EventStart=", format_date(event_start))
  }
  if (!is.null(event_end)) {
    request$url = paste0(request$url, "&EventEnd=", format_date(event_end))
  }
  if (!is.null(publication_from)) {
    request$url = paste0(request$url, "&PublicationFrom=", format_date(publication_from))
  }
  if (!is.null(publication_to)) {
    request$url = paste0(request$url, "&PublicationTo=", format_date(publication_to))
  }
  if (!is.null(participant_id)) {
    request$url = paste0(request$url, "&ParticipantID=", participant_id)
  }
  if (!is.null(asset_id)) {
    request$url = paste0(request$url, "&AssetID=", asset_id)
  }
  if (!is.null(event_type)) {
    request$url = paste0(request$url, "&EventType=", event_type)
  }
  if (!is.null(fuel_type)) {
    request$url = paste0(request$url, "&FuelType=", fuel_type)
  }
  if (!is.null(message_type)) {
    request$url = paste0(request$url, "&MessageType=", message_type)
  }
  if (!is.null(message_id)) {
    request$url = paste0(request$url, "&MessageID=", message_id)
  }
  if (!is.null(unavailability_type)) {
    request$url = paste0(request$url, "&UnavailabilityType=", unavailability_type)
  }
  if (!is.null(active_flag)) {
    request$url = paste0(request$url, "&ActiveFlag=", active_flag)
  }
  if (!is.null(sequence_id)) {
    request$url = paste0(request$url, "&SequenceId=", sequence_id)
  }
  if (!is.null(service_type)){
    request$url = paste0(request$url, "&ServiceType=", service_type)
  }
  request$service_type <- service_type
  request$data_item <- data_item
  return(request)
}


# ================
# my experiments
# ================
# 
  api_key  <- Sys.getenv("scripting_key")
  api_key
  data_item  <- "B1720"
  api_version  <- "v1" 
  SettlementDate  <-  "2019-10-13"
2019-10-13 

  Period="*"

#  request$url <- paste0("https://api.bmreports.com/BMRS/", 
                       data_item, "/", api_version, "?APIKey=", api_key)
                      
  the_url <- paste0("https://api.bmreports.com/BMRS/", 

                       data_item, "/", api_version, 
                       "?APIKey=", api_key, 
                       "&Period=",Period, 
                       "&ServiceType=", "csv",
                       "&SettlementDate=",SettlementDate)
  the_url

#"https://api.bmreports.com/BMRS/B1720/v1?APIKey=4rs1u6b3ror2wjg
&Period=*&SettlementDate=2020-10-10"
  #
res  <- httr::GET(the_url)
# HERE ...
r <- res %>% httr::content()


# more complicated
#
endpoint  <- paste0("https://api.bmreports.com/BMRS/B1720/v1")
settlement_date  <-  "14-12-2016"
settlement_date  <-  "2020-10-10"
api_key  <- api_key

the_url  <- paste0(endpoint,"?APIKey=", api_key, 
                   "&settlement_date=",settlement_date) 


the_url
# Status 200
res  <- httr::GET(the_url)
res


# ERROR
# use params
params  <- list( api_key= api_key,
                settlement_date= settlement_date)
unlist(params)
res  <- httr::GET(endpoint, params)
res


## from his web page, bottom
endpoint  <- paste0("https://api.bmreports.com/BMRS/B1720/v1?")
params  <- list( API_key= api_key,
                Period=27,
                ServiceType="csv",
                SettlementDate= settlement_date)

unlist(params)  
library(magrittr)
res  <- httr::GET(endpoint, query = params)
# res %>% jsonlite::fromJSON()
res %>% httr::content(as="text")

# another one?
# gen_data <- full_request(data_item = "FUELINST",
#                          api_key = api,
#                          from_datetime = "2019-07-01 00:00:00",
#                          to_datetime = "2019-07-03 00:00:00",
#                          service_type = "csv")
# 
params  <- list(API_key= api_key,
                 data_item="FUELINST",
                         from_datetime = "2019-07-01 00:00:00",
                         to_datetime = "2019-07-03 00:00:00",
                         service_type= "csv")
res  <- httr::GET(endpoint, query=params)
res
